internal/mode/static/state/graph/service.go

```diff
@@ -0,0 +1,47 @@
+package graph
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func buildReferencedServices(
+	routes map[types.NamespacedName]*Route,
+) map[types.NamespacedName]struct{} {
+	svcNames := make(map[types.NamespacedName]struct{})
+
+	// routes all have populated ParentRefs from when they were created.
+	//
+	// Get all the service names referenced from all the HTTPRoutes.
+	for _, route := range routes {
+		// If none of the ParentRefs are attached to the Gateway, we want to skip the route.
+		attached := false
+		for _, ref := range route.ParentRefs {
+			if ref.Attachment.Attached {
+				attached = true
+				break
+			}
+		}
+		if !attached {
+			continue
+		}
+
+		if !route.Valid {
+			continue
+		}
+
+		for idx := range route.Rules {
```

Comment:
Hmm, I think it was because I reused another part of existing code that was set up that way, but somewhere along the way I removed the other usage of idx, so changing it to `for _, rule := range route.Rules` makes it a lot simpler!